glib-macros/src/gboxed_derive.rs

```diff
@@ -0,0 +1,127 @@
+        quote! {
+            if ptr.is_null() {
+                None
+            } else {
+                Some(&*(ptr as *mut #name))
+            }
+        }
+    } else {
+        quote! {
+            assert!(!ptr.is_null());
+            Some(&*(ptr as *mut #name))
+        }
+    }
+}
+
+pub fn impl_gboxed(input: &syn::DeriveInput) -> TokenStream {
+    let name = &input.ident;
+
+    let gtype_name = match parse_type_name(&input, "gboxed") {
+        Ok(v) => v,
+        Err(e) => abort_call_site!(
+            "{}: derive(GBoxed) requires #[gboxed(type_name = \"BoxedTypeName\")]",
+            e
+        ),
+    };
+
+    let meta = find_attribute_meta(&input.attrs, "gboxed")
+        .unwrap()
+        .unwrap();
+    let nullable = find_nested_meta(&meta, "nullable").is_some();
+
+    let option_to_ptr = gen_option_to_ptr(nullable);
+    let ptr_to_option = gen_ptr_to_option(name, nullable);
+
+    quote! {
+        impl BoxedType for #name {
+            const NAME: &'static str = #gtype_name;
+
+            fn get_type() -> glib::Type {
+                static mut TYPE_: glib::Type = glib::Type::Invalid;
+                static ONCE: ::std::sync::Once = ::std::sync::Once::new();
+
+                ONCE.call_once(|| {
+                    let type_ = glib::subclass::register_boxed_type::<Self>();
+                    unsafe {
+                        TYPE_ = type_;
+                    }
+                });
+
+                unsafe { TYPE_ }
+            }
+        }
+
+        impl glib::StaticType for #name {
+            fn static_type() -> glib::Type {
+                <#name as glib::subclass::boxed::BoxedType>::get_type()
+            }
+        }
+
+        impl glib::value::SetValue for #name {
+            unsafe fn set_value(value: &mut glib::value::Value, this: &Self) {
+                let ptr: *mut #name = Box::into_raw(Box::new(this.clone()));
+                glib::gobject_sys::g_value_take_boxed(
+                    glib::translate::ToGlibPtrMut::to_glib_none_mut(value).0,
+                    ptr as *mut _,
+                );
+            }
+        }
+
+        impl glib::value::SetValueOptional for #name {
+            unsafe fn set_value_optional(value: &mut glib::value::Value, this: Option<&Self>) {
+                let ptr: *mut #name = #option_to_ptr;
+                glib::gobject_sys::g_value_take_boxed(
+                    glib::translate::ToGlibPtrMut::to_glib_none_mut(value).0,
+                    ptr as *mut _,
+                );
+            }
+        }
+
+        impl<'a> glib::value::FromValueOptional<'a> for &'a #name {
+            unsafe fn from_value_optional(value: &'a glib::value::Value) -> Option<Self> {
+                let ptr = glib::gobject_sys::g_value_get_boxed(
+                    glib::translate::ToGlibPtr::to_glib_none(value).0,
+                );
+                #ptr_to_option
+            }
+        }
+
+        impl<'a> glib::value::FromValue<'a> for &'a #name {
```

Comment:
This one should not be implemented for nullable types

Reply:
fixed